test/controllers/application_controller_test.rb

```diff
@@ -7,15 +7,32 @@ class ApplicationControllerTest < ActionDispatch::IntegrationTest
     @admin_user = create(:admin_user)
   end
 
-  test "should redirect to member loans after login" do
+  test "should redirect to member loans after login if there is no referer" do
     sign_in(@user)
     post user_session_url
     assert_redirected_to member_loans_path
   end
 
-  test "should redirect to admin dashboard after login" do
+  test "should redirect to admin dashboard after login if there is no referer" do
     sign_in(@admin_user)
     post user_session_url
     assert_redirected_to admin_dashboard_path
   end
+
+  test "should redirect to previous page for admin after login if there is a referer" do
+    get new_user_session_path, headers: {referer: "http://www.example.com/original-page"}
+    post user_session_path, params: {user: {email: @admin_user.email, password: "password"}}
+    assert_redirected_to "http://www.example.com/original-page"
+  end
+
+  test "should redirect to previous page for member after login if there is a referer" do
+    # @request.env['HTTP_REFERER'] = 'http://www.example.com/items'
+    sign_in(@user)
+    post user_session_path
+    assert_redirected_to items_path
+  end
+
+  test "ignores a poorly formed referer" do
```

Comment:
I added this test since parsing the referrer can fail if someone sends a request with garbage in it.